packages/@uppy/companion-client/src/Provider.js

```diff
@@ -157,24 +156,33 @@ export default class Provider extends RequestClient {
 
   /** @protected */
   async request (...args) {
+    const authTokenBefore = await this.#getAuthToken()
+
     await this.#refreshingTokenPromise
 
     try {
-      // throw Object.assign(new Error(), { isAuthError: true }) // testing simulate access token expired (to refresh token)
-      // A better way to test this is for example with Google Drive:
+      // to test simulate access token expired (leading to a token token refresh),
+      // see mockAccessTokenExpiredError in companion/drive.
+      // If you want to test refresh token *and* access token invalid, do this for example with Google Drive:
       // While uploading, go to your google account settings,
       // "Third-party apps & services", then click "Companion" and "Remove access".
 
       return await super.request(...args)
     } catch (err) {
       // only handle auth errors (401 from provider), and only handle them if we have a (refresh) token
-      if (!err.isAuthError || !(await this.#getAuthToken())) throw err
+      const authTokenAfter = await this.#getAuthToken()
+      if (!err.isAuthError || !authTokenAfter) throw err
+
+      await this.#refreshingTokenPromise
+
+      const hasAuthTokenAlreadyBeenRefreshed = authTokenBefore !== authTokenAfter
```

Comment:
You're right! It was something I added in total desperation while trying to find the race condition that was caused by `this` changing for every uploaded file (because we created a new Provider instance for every uploaded file). I will remove it